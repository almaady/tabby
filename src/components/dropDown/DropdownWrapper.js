import React from "react";
import PropTypes from "prop-types";
import classNames from "classnames";
import Dropdown from "./Dropdown";
import ButtonIcon from "../buttons/ButtonIcon";
/**
 * @class DropDownWrapper -It contains the Dropdown and the Button Icon that triggers it
 */
class DropdownWrapper extends React.PureComponent {
  /**
   * @property {node} children - The content of the dropdown
   * @property {string} className - Just in case you need another class
   * @property {string} icon - The icon that triggers the dropdown
   * */

  static propTypes = {
    icon: PropTypes.string,
    children: PropTypes.node.isRequired,
    className: PropTypes.string,
    size: PropTypes.string
  };

  constructor(props) {
    super(props);
    this.state = {
      show: false
    };
    this.activatorRef = React.createRef();
    this.dropdownRef = React.createRef();
  }

  componentDidMount() {
    document.addEventListener("mousedown", this.handleClickOutside);
  }

  componentWillUnmount() {
    document.removeEventListener("mousedown", this.handleClickOutside);
  }

  toggleDropDown = () => {
    this.setState(prevState => ({
      show: !prevState.show
    }));
  };

  onClose = () => {
    this.setState({
      show: false
    });
  };

  handleClickOutside = event => {
    const { show } = this.state;
    if (
      this.dropdownRef.current &&
      !this.dropdownRef.current.contains(event.target) &&
      show
    ) {
      this.setState({
        show: false
      });
    }
  };

  render() {
    const { children, className, icon, size } = this.props;

    return (
      <div className={classNames("dropdown__wrapper", className)} ref={this.dropdownRef}>
        <div ref={this.activatorRef} id="activator-dropdown">
          <ButtonIcon icon={icon} onClick={this.toggleDropDown} size={size} />
        </div>
        {this.state.show && 
          <Dropdown onClose={this.onClose} className={className}>
            {children}
          </Dropdown>
        }
      </div>
    );
  }
}

DropdownWrapper.defaultProps = {
  className: "",
  size: "R",
  icon: "options"
};

export default DropdownWrapper;
