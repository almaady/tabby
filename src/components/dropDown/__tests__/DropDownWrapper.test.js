import React from "react";
import { shallow, mount } from "enzyme";
import sinon from "sinon";

import DropdownOption from "../DrodownOption";
import DropdownWrapper from "../DropdownWrapper";

describe("DropDownWrapper", () => {
  const clickOption = sinon.stub()
  let props;
  beforeEach(() => {
    props = {
      children: [
        <DropdownOption onClick={clickOption} text="Option 1" key={1} />,
        <DropdownOption onClick={clickOption} text="Option 2" key={2} />,
        <DropdownOption onClick={clickOption} text="Option 3" key={3} />
      ]
    };
  });
  it("should render correctly with default state", () => {
    const component = shallow(<DropdownWrapper {...props} />);
    expect(component).toMatchSnapshot();
  });
  it("should change state when you click on icon", () => {
    const component = shallow(<DropdownWrapper {...props} />);
    component.find("ButtonIcon").simulate("click");
    expect(component.state().show).toEqual(true);
  });
  it("should chnage state when you click outside", () => {
    const wrapper = mount(
      <div>
        <DropdownWrapper {...props} />
      </div>
    );
    wrapper.find("ButtonIcon").simulate("click");
    const event = new MouseEvent("mousedown");
    document.dispatchEvent(event);
    expect(wrapper.find(DropdownWrapper).state().show).toEqual(false);
  });
});