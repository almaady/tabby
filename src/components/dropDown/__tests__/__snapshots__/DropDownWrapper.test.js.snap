// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropDownWrapper should render correctly with default state 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <DropdownWrapper
    className=""
    icon="options"
    size="R"
  >
    <DropdownOption
      className=""
      icon=""
      iconDirection="left"
      onClick={[Function]}
      text="Option 1"
    />
    <DropdownOption
      className=""
      icon=""
      iconDirection="left"
      onClick={[Function]}
      text="Option 2"
    />
    <DropdownOption
      className=""
      icon=""
      iconDirection="left"
      onClick={[Function]}
      text="Option 3"
    />
  </DropdownWrapper>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          id="activator-dropdown"
        >
          <ButtonIcon
            borderColor=""
            className=""
            color="gray-l1"
            disabled={false}
            icon="options"
            onClick={[Function]}
            size="R"
            type="button"
          />
        </div>,
        false,
      ],
      "className": "dropdown__wrapper",
    },
    "ref": Object {
      "current": null,
    },
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <ButtonIcon
            borderColor=""
            className=""
            color="gray-l1"
            disabled={false}
            icon="options"
            onClick={[Function]}
            size="R"
            type="button"
          />,
          "id": "activator-dropdown",
        },
        "ref": Object {
          "current": null,
        },
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "borderColor": "",
            "className": "",
            "color": "gray-l1",
            "disabled": false,
            "icon": "options",
            "onClick": [Function],
            "size": "R",
            "type": "button",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "div",
      },
      false,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            id="activator-dropdown"
          >
            <ButtonIcon
              borderColor=""
              className=""
              color="gray-l1"
              disabled={false}
              icon="options"
              onClick={[Function]}
              size="R"
              type="button"
            />
          </div>,
          false,
        ],
        "className": "dropdown__wrapper",
      },
      "ref": Object {
        "current": null,
      },
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <ButtonIcon
              borderColor=""
              className=""
              color="gray-l1"
              disabled={false}
              icon="options"
              onClick={[Function]}
              size="R"
              type="button"
            />,
            "id": "activator-dropdown",
          },
          "ref": Object {
            "current": null,
          },
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "borderColor": "",
              "className": "",
              "color": "gray-l1",
              "disabled": false,
              "icon": "options",
              "onClick": [Function],
              "size": "R",
              "type": "button",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
        false,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
